### PR TITLE
[Feat] Add GQA Sink Backward BHSD with robustness improvements

### PR DESCRIPTION
**Pull request by wch0810**
## Description

GQA + Attention Sink Flash Attention Backward (BHSD layout) for Ascend NPU. Single-kernel backward with on-chip intermediates, Developer mode (combineCV, 4 True, zero T.Scope/flag/barrier_all). Responds to ShareableXue review requesting single-kernel bwd architecture.

**2 kernels total** (down from 9 in original, 4 in previous rev):
1. `flashattn_fwd` — Forward (online softmax + sink + window) -> O, lse
2. `flashattn_bwd` — Single kernel: Phase0(Delta) + KV-loop(5 GEMM + softmax + dS) + Phase5(dSink). dK/dV fp32 atomic_add + host .half() cast.

**0 GM workspace in bwd** — all intermediates (S/P/dP/dS/p_delta/ds_delta) on-chip via `alloc_shared`/`alloc_fragment`.

**1 host sync** (bwd → host .half(), waits for atomic_add completion — cannot eliminate, hardware constraint).

Tested on Ascend NPU, fp16, B=1, H=64, N=4096, D=128, groups=8, window=128. Precision verified against PyTorch golden (169-line standard: fp16 atol=6.10e-5, rtol=1.95e-3, L0 7/7 + L1 8/8 PASS). Smoke test checks all 6 outputs (fwd_O/Delta/dQ/dK/dV/dSinks).

PR: https://github.com/tile-ai/tilelang-ascend/pull/1638
Commit: `f943abb`

## Golden Config (B=1, H=64, N=4096, D=128, groups=8, window=128, fp16)

| Kernel | Latency (us) | vs GPU 14287us | Max Diff |
|--------|-------------|----------------|---------|
| TileLang bwd (single flashattn_bwd) | 5127 | 64.1% faster | 3.906e-03 |
| TileLang total (fwd + bwd) | ~7600 | 46.8% faster | — |
| GPU Backward (baseline) | 14287 | 1.00x | — |

- NPU bwd **64.1% faster than GPU** baseline.
- Expert 95% target (4464us) not reached (gap +663us, V→C sync serialization cost).

## Architecture

```
2 kernels:
  1. flashattn_fwd   — Forward (online softmax + sink + window) -> O, lse
  2. flashattn_bwd   — Single kernel bwd:
       Phase 0 (Vector): Delta = sum(O * dO, dim=-1)
       Phase 1-4 (KV loop, C-V-C-V-C per iteration):
         GEMM1 S=Q@K^T → softmax P → GEMM2 dV(+corr) → GEMM3 dP → dS compute → GEMM4 dK(+corr) → GEMM5 dQ
       Phase 5 (Vector): dSink = -exp(sink - lse) * Delta
       dK/dV: fp32 atomic_add + host .half() cast (no postprocess kernel)

Launch chain:
  fwd_mod(q, k, v, sinks)                    ← launch 1
  bwd_mod(q, k, v, dO, O, lse, sinks, ...)   ← launch 2
  torch.npu.synchronize()                    ← host sync 1 (atomic_add completion)
  dK.half(); dV.half()                       ← host D2D cast
```

## Key Technical Points

- **Developer + combineCV (4 True)**: 0 T.Scope/flag/barrier_all/T.mma. combineCV auto-separates Cube/Vector by buffer scope, AUTO_CV_SYNC auto-inserts cross-core sync.
- **alloc_shared/fragment**: Compiler auto-maps to L1/UB/L0C. `T.copy(fragment, shared)` is source-level on-chip direct (compiler auto-inserts GM relay via AUTO_CV_SYNC).
- **0 GM workspace**: All intermediates (S/P/dP/dS/p_delta/ds_delta) on-chip. P_fp32 retained in UB (s_ub) for dS compute.
- **Compensated GEMM**: fp16 GEMM main + fp16 correction (p_delta/ds_delta) via L0C accumulation (init=True main + init=False corr on same L0C).
- **threads=1**: Single thread, no vid.
- **host .half()**: dK/dV fp32 atomic_add + host cast (same pattern as varlen rev3). Eliminates postprocess kernel.

## Why 1 host sync remains (cannot eliminate)

dK/dV receive contributions from multiple Q-blocks via `T.tile.atomic_add` (fp32 GM tensor). Cast to fp16 must wait for all Q-blocks to complete. Single kernel block cannot know if other blocks are done — this is a hardware constraint.

**Attempted fp16 atomic_add** (P3): precision fails. fp16 accumulation rounding error scales with Q-block count: N=256 (4 blocks) ratio~0.97, N=4096 (64 blocks) ratio~0.90. fp32 atomic_add + host .half() is the only viable path.

**Framework-level fix**: `src/op/ascend.cc:753` ICHECK prevents cross-dtype `atomic_add_l0c_to_gm<float, half>` (L0C fp32 → GM fp16). Removing this ICHECK would allow fp32 accumulation with fp16 write (same precision as current), eliminating host .half(). Will raise as issue.

## Evolution (9 → 2 kernels)

| Version | Kernels | bwd launches | GM workspace | host sync | bwd us |
|---------|---------|-------------|-------------|-----------|--------|
| baseline | 9 | 5 | 480MB | 4 | 6152 |
| rev0 combineCV | 6 | 3 | 480MB | 4 | 5020 |
| rev2 single bwd | 4 | 1 | 480MB | 2 | 5556 |
| P1 on-chip | 4 | 1 | 0 | 2 | 5360 |
| P2 merge preprocess | 3 | 1 | 0 | 1 | 5117 |
| **P4 host .half()** | **2** | **1** | **0** | **1** | **5127** |

### CODE REVIEW

File: examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py [hunk 1]
```diff
@@ -0,0 +1,1179 @@
...
+
+            T.copy(dQ[bz, by, bx * blk + vid * blk // 2 : bx * blk + vid * blk // 2 + blk // 2, :], dq_ub)
+            T.copy(dq_ub, dq_half)
+            T.copy(dq_half, dQ_out[bz, by, bx * blk + vid * blk // 2 : bx * blk + vid * blk // 2 + blk // 2, :])
+
+    return main
+
+
+# ============================================================================
+# Kernel 9: Dsink — dSink = -exp(sink - lse) * Delta, fp32 output
```
Comment (hedi515) [inline examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py]: 需要这么多kernel吗，目的是什么

Comment (wch0810) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py]: GPU 的单个 `flashattn_bwd` 反向 Kernel，在 NPU 侧被拆分为 5 个独立 Kernel（k1-k5），主要受昇腾硬件 Cube/Vector 分核约束：

```
k1 (Cube):   S = Q @ K^T                     -> ws_s          [fp32]
k2 (Vector): P = exp(S*scale - lse) + mask    -> ws_p, ws_p_delta, ws_p_fp32
k3 (Cube):   dV = P^T @ dO (Comp GEMM) + dP   -> dV[atomic], ws_dp
k4 (Vector): dS = P*(dP-Delta)*scale + mask   -> ws_ds, ws_ds_delta
k5 (Cube):   dK = dS^T @ Q (Comp GEMM) + dQ   -> dK[atomic], dQ[L0C accumulate]
```

GPU 可以在同一个 Kernel 的流水线循环里，自由交替执行 Tensor-Core GEMM 矩阵计算和 CUDA-Core softmax/mask 向量运算，只用一个 Kernel 就完成 dQ/dK/dV 全部反向梯度。

而昇腾 NPU，矩阵乘（GEMM）跑在 Cube 核，softmax、mask 这类向量运算跑在 Vector 核，Cube、Vector 运算不能混写在同一个 Kernel 内部，必须按运算类型拆成多个 Kernel。

除此之外还有 NPU 特有 workaround：`gemm_v0` fp32 路径存在 codegen bug，k3 的 dV 和 k5 的 dK/dQ 额外引入补偿 GEMM，用 fp16 主项 + 校正项逼近 fp32 精度，GPU 不需要这部分逻辑。

**对应关系**：9 = 4（1:1 相同：fwd / preprocess / postprocess / dsink）+ 5（GPU 1 个 `flashattn_bwd` 拆成 NPU k1-k5）。

Comment (ShareableXue) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py]: you said:
> “GPU 的单个 flashattn_bwd 反向 Kernel，在 NPU 侧被拆分为 5 个独立 Kernel” 

I don't think this statement is correct. Of course Cube and Vector work can be expressed in the same logical TileLang kernel.

They run on different physical subcores and cannot directly share local buffers, but combineCV is precisely the mechanism for connecting Cube and Vector phases through workspace and synchronization within a single T.Kernel.

Comment (ShareableXue) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py]: This PR itself already demonstrates that in flashattn_fwd, flashattn_bwd_qk_softmax, and flashattn_bwd_dv_dp_ds. We also already have an Ascend GQA backward implementation that executes C-V-C-V-C, including all five GEMMs, in one main backward kernel.

Therefore, splitting into five kernels is not a hardware requirement. It is a scheduling/implementation choice in the current version.

Comment (wch0810) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py]: You're right, thanks for the correction. I was under the impression that avoiding T.Scope required Expert mode, and that in Developer mode Cube/Vector work had to be split into separate kernels. I now understand that combineCV is precisely the mechanism to express Cube and Vector phases within a single T.Kernel in Developer mode.
The latest commits on this PR address exactly this — the bwd kernels are merged from 9 → 6 via combineCV, with Cube and Vector phases connected by intra‑kernel workspace and auto‑sync, no T.Scope or manual flag. L0 7/7 + L1 8/8 PASS, bwd main 6152→5020us (-18.5%).

Comment (ShareableXue) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py]: Thanks for explaining the intent. If a bug exists in gemm_v0, we would like to see it raised as an issue and will address it.

**Comment by ShareableXue** [comment]
Thanks for the work on this operator! 
I am requesting changes because I don't think the current multi-kernel decomposition is an appropriate foundation for further incremental review.

The main value of this operator is fusion. The main reason to implement this as a fused attention operator is to keep tiled intermediates on chip and reduce kernel-launch and synchronization overhead.

The current implementation uses five launches for backward, six including forward, spills the intermediate attention tiles to GM, and synchronizes the host between stages. At that point, this is effectively a hand-written composition of small operators. What do we gain over expressing the same computation as a composition of existing PyTorch operators?

Would you please  integrate them into one kernel, instead of treating the current multi-kernel decomposition as the final operator architecture.

File: examples/attention_sink/example_gqa_sink_bwd_bhsd/test_gqa_sink_bwd_bhsd.py [hunk 2]
```diff
@@ -0,0 +1,1651 @@
...
+        print("--- per-kernel breakdown (with syncs) ---")
+        for name, us in per_kernel_us.items():
+            print(f"  {name:20s}: {us:.0f} us")
+
+    print("\nTest Passed!")
+    return True
+
+
+# ============================================================================
+# msprof op: hardware-level kernel profiling (Cube/MTE2/MTE3/L2/Scalar stall)
```
Comment (hedi515) [inline examples/attention_sink/example_gqa_sink_bwd_bhsd/test_gqa_sink_bwd_bhsd.py]: 性能测试不用放进来

**Comment by hedi515** [comment]
lgtm

File: examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py [hunk 3]
```diff
@@ -0,0 +1,606 @@
+"""GQA Sink Attention (BHSD) for Ascend NPU — on-chip single-kernel backward.
```
Comment (LLMZhangYC) [inline examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py:1]: ascendc有无对应实现，性能对比如何

**Comment by wch0810** [comment]
image

File: examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py (hunk 3 above)
Comment (wch0810) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py]: image
是这个吗

File: examples/attention_sink/example_gqa_sink_bwd_bhsd/test_gqa_sink_bwd_bhsd.py (hunk 2 above)
Comment (wch0810) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/test_gqa_sink_bwd_bhsd.py]: Removed. The test file now only contains accuracy tests (L0/L1/L2/Boundary).

File: examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py (hunk 3 above)
Comment (wch0810) [reply examples/attention_sink/example_gqa_sink_bwd_bhsd/example_gqa_sink_bwd_bhsd.py:1]: image

**Review by ChaoyangJi — APPROVED** [review]
approve

**Review by LLMZhangYC — APPROVED** [review]
approve
